### ISSUE TITLE
Weld failure: java.lang.ClassFormatError: Illegal class name 

### ISSUE THREAD

**Issue by arjantijms** [eclipse-ee4j/glassfish#23371]
During deployment when injecting a CDI bean into an EJB.

Code:

```java
@RequestScoped
public class ImplicitTestBean {

}
```

```java
@Stateless
public class SlessEJB implements Sless {

    @Inject
    ImplicitTestBean testBean;
}
```

Exception:

```
java.lang.ClassFormatError: Illegal class name "/Sless$365510969$Proxy$_$$_Weld$EnterpriseProxy$" in class file /Sless$365510969$Proxy$_$$_Weld$EnterpriseProxy$
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1016)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.weld.util.bytecode.ClassFileUtils.toClass2(ClassFileUtils.java:143)
	at org.jboss.weld.util.bytecode.ClassFileUtils.toClass(ClassFileUtils.java:112)
Caught: java.lang.RuntimeException: java.lang.ClassFormatError: Illegal class name "/Sless$365510969$Proxy$_$$_Weld$EnterpriseProxy$" in class file /Sless$365510969$Proxy$_$$_Weld$EnterpriseProxy$
	at org.jboss.weld.util.bytecode.ClassFileUtils.toClass(ClassFileUtils.java:116)
	at org.jboss.weld.bean.proxy.ProxyFactory.createProxyClass(ProxyFactory.java:481)
	at org.jboss.weld.bean.proxy.ProxyFactory.getProxyClass(ProxyFactory.java:370)
Caught: org.jboss.weld.exceptions.WeldException: WELD-001524: Unable to load proxy class for bean Session bean [class SlessEJB with qualifiers [@Any @Default]; local interfaces are [] with class class SlessEJB
	at org.jboss.weld.bean.proxy.ProxyFactory.getProxyClass(ProxyFactory.java:378)
	at org.jboss.weld.module.ejb.SessionBeanProxyInstantiator.<init>(SessionBeanProxyInstantiator.java:48)
	at org.jboss.weld.module.ejb.SessionBeanImpl.initializeAfterBeanDiscovery(SessionBeanImpl.java:283)
	at org.jboss.weld.bootstrap.ConcurrentBeanDeployer$AfterBeanDiscoveryInitializerFactory.doWork(ConcurrentBeanDeployer.java:111)
	at org.jboss.weld.bootstrap.ConcurrentBeanDeployer$AfterBeanDiscoveryInitializerFactory.doWork(ConcurrentBeanDeployer.java:102)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:62)
	at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:55)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)

org.glassfish.deployment.common.DeploymentException: CDI deployment failure:WELD-001524: Unable to load proxy class for bean Session bean [class SlessEJB with qualifiers [@Any @Default]; local interfaces are [] with class class SlessEJB
	at org.glassfish.weld.WeldDeployer.event(WeldDeployer.java:212)
	at org.glassfish.kernel.event.EventsImpl.send(EventsImpl.java:107)
	at org.glassfish.internal.data.ApplicationInfo.load(ApplicationInfo.java:304)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:472)
	at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:195)
	at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:467)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:516)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:512)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:361)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:511)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:542)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:534)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/javax.security.auth.Subject.doAs(Subject.java:361)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:533)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1441)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1823)
	at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1699)
	at org.glassfish.admin.rest.resources.admin.CommandResource.executeCommand(CommandResource.java:384)
	at org.glassfish.admin.rest.resources.admin.CommandResource.execCommandSimpInMultOut(CommandResource.java:211)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:475)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:397)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
	at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:356)
	at org.glassfish.admin.rest.adapter.JerseyContainerCommandService$3.service(JerseyContainerCommandService.java:150)
	at org.glassfish.admin.rest.adapter.RestAdapter.service(RestAdapter.java:155)
	at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:440)
	at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:144)
	at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:174)
	at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:153)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:196)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:88)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:246)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:178)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:118)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:96)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:51)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:510)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:82)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:83)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:34)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:101)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:535)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:515)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by arjantijms** [comment]
The error appeared to be unrelated to the CDI bean injection, but having to do with the bean being in the default package.

Created Weld issue here: https://issues.redhat.com/browse/WELD-2657

**Comment by manovotn** [comment]
In case of default package, which I should say is a) bad practice and b) a corner case, Weld will have to use a package of its own making to define the proxy into as you are not allowed to define classes into default package. This package is (with 4.0.0.Final) something along the lines of `org.jboss.weld.proxies`. I am not sure if we have this tested automatically (as all tests are in actual packages) so I will check if I can reproduce it. It shouldn't really be an issue only with EJB though because the mechanism for proxy definition is shared.

Another point - from the log I can see that it's using `ClassFileUtils` to define a new class. All integrators should provide an implementation of [`ProxyServices`](https://github.com/weld/api/blob/master/weld-spi/src/main/java/org/jboss/weld/serialization/spi/ProxyServices.java) which allows Weld to let the integrator define classes in whatever way they need to handle it. This will also allow them to avoid issues with illegal access error from JDK 11 as Weld 4.0.0.Final attempts to crack open `ClassLoader` methods.

The interesting part is the JDK 8 working and JDK 11 failing. That almost sounds like there are different limitations in regards to class defining.

**Comment by arjantijms** [comment]
> In case of default package, which I should say is a) bad practice and b) a corner case

I'm definitely not a fan of that, but the existing tests just happened to use that. Maybe the TCK in one of its million+ tests uses it somewhere as well (we've not yet come to the point of running the TCK on GF 6.1).

Thanks for looking into it. I'll take a look at ProxyServices then, you mentioned it before. Thanks!

**Comment by arjantijms** [comment]
@manovotn 
>  It shouldn't really be an issue only with EJB though because the mechanism for proxy definition is shared.

True, there's more failures, e.g. for this non-EJB class which is in the default package as well. Someone back then must have really liked this non-package:

```java
@jakarta.enterprise.context.RequestScoped
public class Bar{
@jakarta.inject.Inject
Foo f;
}
```

(that same person must also have liked obscure formatting, but that aside)

This one results in:

```
java.lang.ClassFormatError: Illegal class name "/Bar$Proxy$_$$_WeldClientProxy" in class file /Bar$Proxy$_$$_WeldClientProxy
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1016)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.weld.util.bytecode.ClassFileUtils.toClass2(ClassFileUtils.java:143)
	at org.jboss.weld.util.bytecode.ClassFileUtils.toClass(ClassFileUtils.java:112)
	at org.jboss.weld.bean.proxy.ProxyFactory.createProxyClass(ProxyFactory.java:481)
	at org.jboss.weld.bean.proxy.ProxyFactory.getProxyClass(ProxyFactory.java:370)
	at org.jboss.weld.bean.proxy.ProxyFactory.run(ProxyFactory.java:331)
	at org.jboss.weld.bean.proxy.ProxyFactory.create(ProxyFactory.java:323)
	at org.jboss.weld.bean.proxy.ClientProxyFactory.create(ClientProxyFactory.java:83)
	at org.jboss.weld.bean.proxy.ClientProxyProvider.createClientProxy(ClientProxyProvider.java:205)
	at org.jboss.weld.bean.proxy.ClientProxyProvider.createClientProxy(ClientProxyProvider.java:195)
	at org.jboss.weld.bean.proxy.ClientProxyProvider.access$100(ClientProxyProvider.java:44)
	at org.jboss.weld.bean.proxy.ClientProxyProvider$CreateClientProxy.apply(ClientProxyProvider.java:52)
	at org.jboss.weld.bean.proxy.ClientProxyProvider$CreateClientProxy.apply(ClientProxyProvider.java:48)
	at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.lambda$new$0(ReentrantMapBackedComputingCache.java:55)
	at org.jboss.weld.util.LazyValueHolder$1.computeValue(LazyValueHolder.java:32)
	at org.jboss.weld.util.LazyValueHolder.get(LazyValueHolder.java:46)
	at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.getValue(ReentrantMapBackedComputingCache.java:72)
	at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.getCastValue(ReentrantMapBackedComputingCache.java:78)
	at org.jboss.weld.bean.proxy.ClientProxyProvider.getClientProxy(ClientProxyProvider.java:229)
	at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:688)
	at org.jboss.weld.manager.BeanManagerImpl.getInjectableReference(BeanManagerImpl.java:794)
	at org.jboss.weld.injection.FieldInjectionPoint.inject(FieldInjectionPoint.java:92)
	at org.jboss.weld.util.Beans.injectBoundFields(Beans.java:336)
	at org.jboss.weld.util.Beans.injectFieldsAndInitializers(Beans.java:347)
	at org.jboss.weld.injection.producer.ResourceInjector$1.proceed(ResourceInjector.java:69)
	at org.glassfish.weld.services.InjectionServicesImpl.aroundInject(InjectionServicesImpl.java:149)
	at org.jboss.weld.injection.InjectionContextImpl.run(InjectionContextImpl.java:46)
	at org.jboss.weld.injection.producer.ResourceInjector.inject(ResourceInjector.java:71)
	at org.jboss.weld.injection.producer.BasicInjectionTarget.inject(BasicInjectionTarget.java:117)
	at org.glassfish.weld.services.JCDIServiceImpl.createManagedObject(JCDIServiceImpl.java:456)
	at org.glassfish.weld.services.JCDIServiceImpl.createManagedObject(JCDIServiceImpl.java:312)
	at com.sun.enterprise.container.common.impl.managedbean.ManagedBeanManagerImpl.createManagedBean(ManagedBeanManagerImpl.java:453)
	at com.sun.enterprise.container.common.impl.managedbean.ManagedBeanManagerImpl.createManagedBean(ManagedBeanManagerImpl.java:406)
	at com.sun.enterprise.container.common.impl.util.InjectionManagerImpl.createManagedObject(InjectionManagerImpl.java:312)
	at com.sun.enterprise.web.WebContainer.createServletInstance(WebContainer.java:725)
	at com.sun.enterprise.web.WebModule.createServletInstance(WebModule.java:1946)
	at org.apache.catalina.core.StandardWrapper.loadServlet(StandardWrapper.java:1262)
	at org.apache.catalina.core.StandardWrapper.allocate(StandardWrapper.java:1069)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:136)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:119)
	at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:611)
	at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:550)
```

**Comment by manovotn** [comment]
I didn't have much time today so I'll get to this on Mon.
But I did manage to reproduce this purely in SE. It's going to be some issue with how we handle default package and probably a regression if you say it worked before.
Looks like we don't even test this case because of how such test would have to be structured so this probably slipped us somewhere along the way :-/

**Comment by manovotn** [comment]
Ok, I found the source of the problem. It is actually not any change in Weld but a change in how JDK perceives empty packages.
JDK 8 javadoc for `Class.getPackage` will return `null` for empty package whereas JDK will instead return what looks like valid package but with empty name (which is actually an `unnamed package` in terms of JDK).
Weld is simply expecting the former and doesn't otherwise use our custom package. This should be an easy fix; PRs will be linked to Weld JIRA in case you wish to follow that.

JDK 8 docs - [link](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html#getPackage--)
JDK 11 docs - [link](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Class.html#getPackage())

@arjantijms is this something that's blocking testing and needs a Weld release or are you fixing this on TCK side?

**Comment by pzygielo** [comment]
> Ok, I found the source of the problem. It is actually not any change in Weld but a change in how JDK perceives empty packages.

Nice finding. Just for completeness - it was announced in https://www.oracle.com/java/technologies/javase/9-relnotes.html first.

**Comment by manovotn** [comment]
This is now fixed in Weld and will be shipped in `4.0.1.Final` and `3.1.7.Final`.

**Comment by arjantijms** [comment]
Thank you again @manovotn! We'll update to `4.0.1.Final` as soon as its released!

**Comment by arjantijms** [comment]
> @arjantijms is this something that's blocking testing and needs a Weld release or are you fixing this on TCK side?

I've worked around the blocking tests by moving the classes to a named package. As it concerned only 3 tests (for now) this was doable.

**Comment by arjantijms** [comment]
p.s.

> All integrators should provide an implementation of ProxyServices 

At a glance, I noticed we actually have one. See 

https://github.com/eclipse-ee4j/glassfish/blob/master/appserver/web/weld-integration/src/main/java/org/glassfish/weld/services/ProxyServicesImpl.java

Maybe it's not actually used, have to do some digging for that.

**Comment by manovotn** [comment]
Even if it were used, it doesn't implement all methods - the ones for class defining are actually defaulted.
And I am thinking that with the latest changes (for JDK 11 friendly class defining by default) this would actually break you if you were using it because Weld would detect an implementation and would presume it to be able to handle proxy definition. But I'd need to double check that.

**Comment by arjantijms** [comment]
Fixed iin 6.2.
